theseus/geometry/utils/utils.py

```diff
@@ -0,0 +1,92 @@
+# Copyright (c) Meta Platforms, Inc. and affiliates.
+#
+# This source code is licensed under the MIT license found in the
+# LICENSE file in the root directory of this source tree.
+
+import threading
+from typing import Any
+
+import torch
+
+
+class LieGroupContext(object):
+    contexts = threading.local()
+
+    @classmethod
+    def get_context(cls):
+        if not hasattr(cls.contexts, "use_lie_tangent"):
+            cls.contexts.use_lie_tangent = False
+        return cls.contexts.use_lie_tangent
+
+    @classmethod
+    def set_context(cls, use_lie_tangent: bool):
+        cls.contexts.use_lie_tangent = use_lie_tangent
+
+
+class set_lie_tangent_enabled(object):
+    def __init__(self, mode: bool) -> None:
+        self.prev = LieGroupContext.get_context()
+        LieGroupContext.set_context(mode)
+
+    def __enter__(self) -> None:
+        pass
+
+    def __exit__(self, exc_type: Any, exc_value: Any, traceback: Any) -> None:
+        LieGroupContext.set_context(self.prev)
+
+
+class enable_lie_tangent(object):
+    def __enter__(self) -> None:
+        self.prev = LieGroupContext.get_context()
+        LieGroupContext.set_context(True)
+
+    def __exit__(self, typ, value, traceback) -> None:
+        LieGroupContext.set_context(self.prev)
+
+
+class no_lie_tangent(LieGroupContext):
+    def __enter__(self):
+        self.prev = super().get_context()
+        LieGroupContext.set_context(False)
+        return self
+
+    def __exit__(self, typ, value, traceback):
+        LieGroupContext.set_context(self.prev)
+
+
+class LieGroupTensor(torch.Tensor):
+    from torch._C import _disabled_torch_function_impl
```

Comment:
Is this related to [this](https://github.com/pytorch/pytorch/issues/46159#issuecomment-707817037) issue? Can you explain a bit why this part is necessary?